src/storage-controller/src/lib.rs

```diff
@@ -797,17 +797,59 @@ where
                         "cannot have multiple IDs for introspection type"
                     );
 
-                    self.collection_manager.register_collection(id);
+                    let metadata = self.storage_collections.collection_metadata(id)?.clone();
+                    let persist_client = self
+                        .persist
+                        .open(metadata.persist_location.clone())
+                        .await
+                        .unwrap();
+
+                    let read_handle_fn = move || {
+                        let persist_client = persist_client.clone();
+                        let metadata = metadata.clone();
+
+                        let fut = async move {
+                            let read_handle = persist_client
+                                .open_leased_reader::<SourceData, (), T, Diff>(
+                                    metadata.data_shard,
+                                    Arc::new(metadata.relation_desc.clone()),
+                                    Arc::new(UnitSchema),
+                                    Diagnostics {
+                                        shard_name: id.to_string(),
+                                        handle_purpose: format!("snapshot {}", id),
+                                    },
+                                    USE_CRITICAL_SINCE_SNAPSHOT.get(persist_client.dyncfgs()),
+                                )
+                                .await
+                                .expect("invalid persist usage");
+                            read_handle
+                        };
+
+                        fut.boxed()
+                    };
+
+                    // How to decide between collection types:
+                    //
+                    // Collection at any given time is derived from some state
+                    // at that time. In a way, it's a view on top of something
+                    // -> incremental collection
+                    //
+                    // Collection records things that happen over time, like a
+                    // history of something -> append-only collection.
```

Comment:
This comment was more confusing to me than it helped, tbh. Isn't the distinction just "updates contain retractions" vs "updates don't contain retractions"?